src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,824 @@
+{
+    struct lyd_node *node = NULL;
+    const struct ly_ctx *ctx = NULL;
+    struct lys_module *module = NULL;
+    char *user = NULL, *group = NULL, *path = NULL, *module_name = NULL;
+    mode_t perm;
+    DIR *dir = NULL;
+    struct dirent *dirent = NULL;
+
+    VRB("Confirmed commit timeout reached. Restoring original running.");
+    if (sr_session_switch_ds(np2srv.sr_sess, SR_DS_RUNNING)) {
+        ERR("Failed switching session.");
+        goto cleanup;
+    }
+    ctx = sr_get_context(np2srv.sr_conn);
+
+    /* Iterate over all files in backup directory */
+    dir = opendir(NP2SRV_DIR);
+    if (!dir) {
+        ERR("Failed opening netopeer2 server directory '%s'", NP2SRV_DIR);
+        goto cleanup;
+    }
+    while ((dirent = readdir(dir))) {
+        if (!strcmp(dirent->d_name, ".") || !strcmp(dirent->d_name, "..") || !strcmp(dirent->d_name, "restore")) {
+            continue;
+        }
+
+        /* try to find the module that corresponds with the file */
+        free(module_name);
+        module_name = NULL;
+        module_name = get_module_name_from_filename(dirent->d_name);
+        if (!module_name) {
+            ERR("Unexpected file '%s'", dirent->d_name);
+            goto cleanup;
+        }
+        module = ly_ctx_get_module_implemented(ctx, module_name);
+        if (!module) {
+            ERR("Module '%s' not implemented", module_name);
+        }
+
+        /* check if module should and can be restored up */
+        if (!np_ly_mod_has_data(module, LYS_CONFIG_W)) {
+            continue;
+        }
+        free(user);
+        user = NULL;
+        free(group);
+        group = NULL;
+        sr_get_module_access(np2srv.sr_conn, module->name, &user, &group, &perm);
+        if (check_permissions(user, group, perm, W)) {
+            continue;
+        }
+
+        /* get, restore and delete the backup */
+        VRB("Rolling back module: %s", module->name);
+        free(path);
+        path = NULL;
+        if (asprintf(&path, "%s/%s", NP2SRV_DIR, dirent->d_name) == -1) {
+            goto cleanup;
+        }
+        if ((get_running_backup(ctx, path, &node))) {
+            rename_failed_file(module_name, path);
+            continue;
+        }
+        if ((sr_replace_config(np2srv.sr_sess, module->name, node, np2srv.sr_timeout))) {
+            ERR("Failed restoring backup for module: '%s'.", module->name);
+            rename_failed_file(module_name, path);
+            continue;
+        }
+        if (unlink(path) == -1) {
+            ERR("Failed removing backup file '%s'", path);
+            goto cleanup;
+        }
+    }
+
+    if (unlink(NP2SRV_DIR RESTORE_FILE) == -1) {
+        ERR("Failed removing confirmed commit restore file.");
+    }
+
+cleanup:
+    free(user);
+    free(group);
+    free(path);
+    free(module_name);
+}
+
+/**
+ * @brief Skip the timer. Call the timer callback function instantly. Disarm the timer.
+ */
+void
+ncc_skip_timer(void)
```

Comment:
Vague function name, I had to read the description to understand. Something like `ncc_trigger_timer` would be self-explaining.